build/teamcity/util/roachtest_util.sh

```diff
@@ -62,8 +65,30 @@ function upload_stats {
   fi
 }
 
-# Upload any stats.json we can find, no matter what happens.
-trap upload_stats EXIT
+set -x
+
+# Uploads roachprod and roachtest binaries to GCS.
+function upload_binaries {
+  if tc_release_branch; then
+      bucket="cockroach-nightly"
+      branch=$(tc_build_branch)
+      arch=$(get_host_arch)
+      os=linux
+      sha=${BUILD_VCS_NUMBER}
+      gsutil cp bin/roachprod gs://$bucket/binaries/$branch/$os/$arch/roachprod.$sha
```

Comment:
**Optional**: Would it be better if we instead just had one file `$os/$arch/roachprod.$branch.latest`? Or, even just one `latest` binary for `master` (per OS/arch combo)?

We only need the latest one, right? Realistically, people are not going to be downloading `roachprod` for older SHA's. This would also obviate the need for garbage collection on the bucket.

Reply:
> Or, even just one latest binary for master (per OS/arch combo)?

That was my original thought too. However, I opted for keeping nightly builds in case of regressions.

>  Realistically, people are not going to be downloading roachprod for older SHA's.

In most cases, that's true. When we do encounter a regression, it would be nice to tell folks, grab the binary with this SHA or wait until a fix is merged. The latter may take time, so it's nice to have a quick "workaround".

Reply:
Added a retention policy for binaries,

<img width="1920" alt="retention_policy" src="https://github.com/cockroachdb/cockroach/assets/1616764/50b2fe52-f9b2-487a-94fb-4d7a17462d47">